Constantly running into 502 errors

I have set up tubesync to connect to an external database to improve performance and yet, when I am adding/deleting larger channels, I always get "502 bad gateway" errors.

Changing a single parameter fixes this issue and makes it work just fine: 
https://github.com/meeb/tubesync/blob/b11b667affcefaeb5eea8c762ba78f8774a84a4c/tubesync/tubesync/gunicorn.py#L25

Can this be updated to a more sensible value?

In my logs i see:

```
10.89.0.2 - - [30/Oct/2024:10:54:08 +0100] "GET /tubesync/sources HTTP/1.1" 200 2446 "https://yt.xxx.org/tubesync/" "Mozilla/5.0 (X11; Linux x86_64; rv:131.0) Gecko/20100101 Firefox/131.0"
[2024-10-30 13:01:50 +0100] [326] [CRITICAL] WORKER TIMEOUT (pid:396)
10.89.0.2 - - [30/Oct/2024:13:01:51 +0100] "GET /tubesync/ HTTP/1.1" 502 150 "https://yt.xxx.org/" "Mozilla/5.0 (X11; Linux x86_64; rv:131.0) Gecko/20100101 Firefox/131.0"
2024/10/30 13:01:51 [error] 365#365: *89 upstream prematurely closed connection while reading response header from upstream, client: 10.89.0.2, server: _, request: "GET /tubesync/ HTTP/1.1", upstream: "http://127.0.0.1:8080/tubesync/", host: "yt.xxx.org", referrer: "https://yt.xxx.org/"
[2024-10-30 13:01:51 +0100] [326] [ERROR] Worker (pid:396) was sent SIGKILL! Perhaps out of memory?
```

`[2024-10-30 13:01:51 +0100] [326] [ERROR] Worker (pid:396) was sent SIGKILL! Perhaps out of memory?`

Basically, something on your system killed a worker process which is why you got a 502. The most common is the OOM killing it, but there are any number of reasons your system might kill a process. Check your host logs etc.

--- a/config/root/etc/nginx/nginx.conf
+++ b/config/root/etc/nginx/nginx.conf
@@ -103,15 +103,16 @@ http {
         # Authentication and proxying
         location / {
             proxy_pass http://127.0.0.1:8080;
-            proxy_set_header Host localhost:8080;
+            proxy_set_header Host localhost:$proxy_port;
             proxy_set_header X-Forwarded-Host $x_forwarded_host;
             proxy_set_header X-Forwarded-Port $x_forwarded_port;
             proxy_set_header X-Forwarded-Proto $x_forwarded_proto;
             proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
             proxy_set_header X-Real-IP $remote_addr;
             proxy_redirect off;
-            proxy_read_timeout 59;
-            proxy_connect_timeout 10;
+            # this read timeout should be lower than gunicorn's timeout
+            proxy_read_timeout 89s;
+            proxy_connect_timeout 10s;
         }
 
         # File dwnload and streaming

--- a/config/root/etc/nginx/nginx.conf
+++ b/config/root/etc/nginx/nginx.conf
@@ -50,6 +50,14 @@ http {
     gzip_http_version 1.1;
     gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;
 
+    # Caching proxy requests
+    proxy_cache_lock on;
+    proxy_cache_use_stale updating;
+    # temporary files in shared memory
+    proxy_temp_path /dev/shm/nginx/tmp 1;
+    # change this to /config/cache/nginx for a persistent cache
+    proxy_cache_path /dev/shm/nginx/cache levels=1:2:2 keys_zone=gunicorn:4m inactive=48h max_size=256m min_free=16m;
+    
     # X-Forwarded-Host (pass-through, or set)
     map $http_x_forwarded_host $x_forwarded_host {
         default $http_x_forwarded_host;
@@ -110,9 +118,14 @@ http {
             proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
             proxy_set_header X-Real-IP $remote_addr;
             proxy_redirect off;
+
             # this read timeout should be lower than gunicorn's timeout
             proxy_read_timeout 89s;
             proxy_connect_timeout 10s;
+
+            # cache long running web requests
+            proxy_cache gunicorn;
+            proxy_cache_lock_timeout 88s;
         }
 
         # File dwnload and streaming

--- a/config/root/etc/nginx/nginx.conf
+++ b/config/root/etc/nginx/nginx.conf
@@ -54,9 +54,9 @@ http {
     proxy_cache_lock on;
     proxy_cache_use_stale updating;
     # temporary files in shared memory
-    proxy_temp_path /dev/shm/nginx/tmp 1;
+    proxy_temp_path /dev/shm/nginx-tmp 1;
     # change this to /config/cache/nginx for a persistent cache
-    proxy_cache_path /dev/shm/nginx/cache levels=1:2:2 keys_zone=gunicorn:4m inactive=48h max_size=256m min_free=16m;
+    proxy_cache_path /dev/shm/nginx-cache levels=1:2:2 keys_zone=gunicorn:4m inactive=48h max_size=256m min_free=16m;
 
     # X-Forwarded-Host (pass-through, or set)
     map $http_x_forwarded_host $x_forwarded_host {

--- a/config/root/etc/nginx/nginx.conf
+++ b/config/root/etc/nginx/nginx.conf
@@ -6,49 +6,49 @@ worker_cpu_affinity auto;
 pid /run/nginx.pid;
 
 events {
-  worker_connections 1024;
+    worker_connections 1024;
 }
 
 http {
 
-  # Basic settings
-  sendfile on;
-  tcp_nopush on;
-  tcp_nodelay on;
-  keepalive_timeout 300;
-  types_hash_max_size 2048;
-  server_tokens off;
-  server_names_hash_bucket_size 64;
-  server_name_in_redirect off;
-  client_body_in_file_only clean;
-  client_body_buffer_size 32K;
-  client_max_body_size 100M;
-  send_timeout 300s;
-  large_client_header_buffers 4 8k;
+    # Basic settings
+    sendfile on;
+    tcp_nopush on;
+    tcp_nodelay on;
+    keepalive_timeout 300;
+    types_hash_max_size 2048;
+    server_tokens off;
+    server_names_hash_bucket_size 64;
+    server_name_in_redirect off;
+    client_body_in_file_only clean;
+    client_body_buffer_size 32K;
+    client_max_body_size 100M;
+    send_timeout 300s;
+    large_client_header_buffers 4 8k;
 
-  # Mime type handling
-  include /etc/nginx/mime.types;
-  default_type application/octet-stream;
+    # Mime type handling
+    include /etc/nginx/mime.types;
+    default_type application/octet-stream;
 
-  # Default security headers
-  add_header X-Frame-Options SAMEORIGIN;
-  add_header X-Content-Type-Options nosniff;
-  add_header X-XSS-Protection "1; mode=block";
+    # Default security headers
+    add_header X-Frame-Options SAMEORIGIN;
+    add_header X-Content-Type-Options nosniff;
+    add_header X-XSS-Protection "1; mode=block";
 
-  # Logging
-  log_format host '$remote_addr - $remote_user [$time_local] "[$host] $request" $status $bytes_sent "$http_referer" "$http_user_agent" "$gzip_ratio"';
-  access_log /dev/stdout;
-  error_log stderr;
+    # Logging
+    log_format host '$remote_addr - $remote_user [$time_local] "[$host] $request" $status $bytes_sent "$http_referer" "$http_user_agent" "$gzip_ratio"';
+    access_log /dev/stdout;
+    error_log stderr;
 
-  # GZIP
-  gzip on;
-  gzip_disable "msie6";
-  gzip_vary on;
-  gzip_proxied any;
-  gzip_comp_level 6;
-  gzip_buffers 16 8k;
-  gzip_http_version 1.1;
-  gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;
+    # GZIP
+    gzip on;
+    gzip_disable "msie6";
+    gzip_vary on;
+    gzip_proxied any;
+    gzip_comp_level 6;
+    gzip_buffers 16 8k;
+    gzip_http_version 1.1;
+    gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;
 
     # X-Forwarded-Host (pass-through, or set)
     map $http_x_forwarded_host $x_forwarded_host {

--- a/config/root/etc/nginx/nginx.conf
+++ b/config/root/etc/nginx/nginx.conf
@@ -57,7 +57,7 @@ http {
     proxy_temp_path /dev/shm/nginx/tmp 1;
     # change this to /config/cache/nginx for a persistent cache
     proxy_cache_path /dev/shm/nginx/cache levels=1:2:2 keys_zone=gunicorn:4m inactive=48h max_size=256m min_free=16m;
-    
+
     # X-Forwarded-Host (pass-through, or set)
     map $http_x_forwarded_host $x_forwarded_host {
         default $http_x_forwarded_host;

--- a/tubesync/tubesync/gunicorn.py
+++ b/tubesync/tubesync/gunicorn.py
@@ -23,7 +23,7 @@ def get_bind():
 
 
 workers = get_num_workers()
-timeout = 30
+timeout = 90
 chdir = '/app'
 daemon = False
 pidfile = '/run/app/gunicorn.pid'

